internal/registry/var.go

```diff
@@ -0,0 +1,126 @@
+package registry
+
+import (
+	"go/types"
+	"strings"
+)
+
+// Var represents a method variable/parameter.
+//
+// It should be created using a method scope instance.
+type Var struct {
+	vr         *types.Var
+	imports    map[string]*Package
+	moqPkgPath string
+
+	Name string
+}
+
+// IsSlice returns whether the type (or the underlying type) is a slice.
+func (v Var) IsSlice() bool {
+	_, ok := v.vr.Type().Underlying().(*types.Slice)
+	return ok
+}
+
+// TypeString returns the variable type with the package qualifier in the
+// format 'pkg.Type'.
+func (v Var) TypeString() string {
+	return types.TypeString(v.vr.Type(), v.packageQualifier)
+}
+
+// packageQualifier is a types.Qualifier.
+func (v Var) packageQualifier(pkg *types.Package) string {
+	path := stripVendorPath(pkg.Path())
+	if v.moqPkgPath != "" && v.moqPkgPath == path {
+		return ""
+	}
+
+	return v.imports[path].Qualifier()
+}
+
+// generateVarName generates a name for the variable using the type
+// information.
+//
+// Examples:
+// - string -> s
+// - int -> n
+// - chan int -> intCh
+// - []a.MyType -> myTypes
+// - map[string]int -> stringToInt
+// - error -> err
+// - a.MyType -> myType
+func generateVarName(t types.Type) string {
+	if t, ok := t.(*types.Named); ok {
+		name := deCapitalise(t.Obj().Name())
+		if name == "error" {
+			name = "err"
+		} else if name == t.Obj().Name() {
+			return name + "MoqParam"
+		}
+
+		return name
+	}
+
+	nestedType := func(t types.Type) string {
+		switch t := t.(type) {
+		case *types.Basic:
+			return t.String()
+
+		case *types.Named:
+			return t.Obj().Name()
+		}
+		return generateVarName(t)
+	}
+
+	switch t := t.(type) {
+	case *types.Basic:
+		return basicTypeVarName(t)
+
+	case *types.Array:
+		return deCapitalise(nestedType(t.Elem())) + "s"
+
+	case *types.Slice:
+		return deCapitalise(nestedType(t.Elem())) + "s"
+
+	case *types.Struct: // anonymous struct
+		return "val"
+
+	case *types.Pointer:
+		return deCapitalise(nestedType(t.Elem()))
```

Comment:
I noticed that this can shadow a type. For example:

```go
type ball struct {}
type sport interface {
        throw(*ball)
}
```
will produce:
```go
// throw calls throwFunc.
func (mock *sportMock) throw(ball *ball) {
	if mock.throwFunc == nil {
		panic("sportMock.throwFunc: method is nil but sport.throw was just called")
	}
	callInfo := struct {
		Ball *ball
	}{
		Ball: ball,
	}
	mock.lockthrow.Lock()
	mock.calls.throw = append(mock.calls.throw, callInfo)
	mock.lockthrow.Unlock()
	mock.throwFunc(ball)
}
```

Reply:
Thanks @pehowell, for pointing that out. I have pushed a commit (6a27d0b) which fixes this and updates the test to cover the scenario.